File: app/src/main/java/com/barryyang/barryyangdemo/java/model/Person.java
package com.barryyang.barryyangdemo.java.model;

import android.os.Build;

import java.util.Objects;

/**
 * @author : BarryYang
 * @date : 2020/7/29 11:20 AM
 * @desc :
 */
public class Person {

    private long userId;
    private String username;

    public Person(long userId, String username) {
        this.userId = userId;
        this.username = username;
    }

    public long getUserId() {
        return userId;
    }

    public void setUserId(long userId) {
        this.userId = userId;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    @Override
    public boolean equals(Object object) {
        if (object instanceof Person) {
            Person person = (Person) object;
            return userId == person.userId && username.equalsIgnoreCase(person.username);
        }
        return false;
    }

    @Override
    public int hashCode() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            return Objects.hash(userId, username);
        }
        return Long.valueOf(userId).hashCode() + username.hashCode();
    }
}

File: app/src/main/java/com/barryyang/barryyangdemo/kotlin/constructor/Person.kt
package com.barryyang.barryyangdemo.kotlin.constructor

/**
 * @author : BarryYang
 * @date : 2020/8/20 3:36 PM
 * @desc :
 */
class Person(val name: String) {

    val children = mutableListOf<Person>();

    /**
     * 如果类有一个主构造函数，每个次构造函数需要委托主构造函数，可以直接委托或者通过别的次构造函数间接
     * 委托，委托到同一个类的另一个构造函数用this关键字即可。
     */
    constructor(name: String, person: Person) : this(name) {
        person.children.add(this)
    }
}